packages/@aws-cdk/core/lib/private/uniqueid.ts

```diff
@@ -20,16 +20,40 @@ const HASH_LEN = 8;
 const MAX_HUMAN_LEN = 240; // max ID len is 255
 const MAX_ID_LEN = 255;
 
+function findLongestDuplicatedSubstring(str: string) {
+  let j = 1;
+  let start = 0;
+  let end = -1;
+
+  for (let i = 0; i < str.length; i++) {
+    let longest = str.slice(i, i + j);
+    const rest = str.slice(i+1);
+
+    while (rest.includes(longest)) {
+      j += 1;
+      longest = str.slice(i, i+j);
+      start = i;
+      end = i + j - 1;
+    }
+  }
+
+  return {
+    start,
+    end,
+  };
+}
+
 /**
  * Calculates a unique ID for a set of textual components.
  *
  * This is done by calculating a hash on the full path and using it as a suffix
  * of a length-limited "human" rendition of the path components.
  *
  * @param components The path components
- * @returns a unique alpha-numeric identifier with a maximum length of 255
+ * @param maxLength The maximum length of the return value
+ * @returns a unique alpha-numeric identifier with a maximum length of 255 or the maxLength parameter
  */
-export function makeUniqueId(components: string[]) {
+export function makeUniqueId(components: string[], maxLength?: number) {
```

Comment:
This [function](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/core/lib/private/unique-resource-name.ts#L50) is already doing what these changes aim to do. I think we should use this instead. The only places where `makeUniqueId` is currently used is in `stack.ts`, where you're making the change and in `Names.uniqueId` which is not actually a unique id and should be deprecated. Can we use `makeUniqueResourceName` instead so we can eliminate the nearly duplicated code?